### ISSUE TITLE
no-unused-vars incorrectly matching destructured function arguments.

### ISSUE THREAD

**Issue by phyllisstein** [eslint/eslint#9493]
Hey folks! I'm having some trouble with the `no-unused-vars` rule when destructuring function arguments. Even variables that should match the ignore pattern are being flagged as unused by ESLint. I've tried tweaking everything I can think to tweak to no avail---would be grateful for any insight you can provide! Some details are below, but please let me know if there's more debugging info I can furnish. Thanks a bunch!

---

**Tell us about your environment**

* **ESLint Version:** v4.9.0
* **Node Version:** v8.7.0
* **npm Version:** v5.5.1

**What parser (default, Babel-ESLint, etc.) are you using?** `babel-eslint@8.0.1`

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```yaml
env:
  browser: true
  es6: true
  node: true
parser: babel-eslint
parserOptions:
  ecmaFeatures:
    experimentalObjectRestSpread: true
    impliedStrict: true
    jsx: true
  ecmaVersion: 8
  sourceType: module
plugins:
  - babel
rules:
  no-unused-vars:
    - error
    - args: after-used
      ignoreRestSiblings: false
      vars: all
      varsIgnorePattern: '^_.+'
```

</details>

**What did you do? Please include the actual source code causing the issue.**

```js
const fun = ({ foo, bar: _bar }) => console.log(foo)
fun()
```

**What did you expect to happen?**
I expected neither `foo` (which is used in the function) nor `_bar` (which is a match for the `varsIgnorePattern`) to be flagged as unused.

**What actually happened? Please include the actual, raw output from ESLint.**
```
/Users/daniel/Repos/Ignota/disibodenberg/test.js
  1:26  error  '_bar' is defined but never used. Allowed unused vars must match /^_.+/  no-unused-vars

✖ 1 problem (1 error, 0 warnings)
```

**Comment by i-ron-y** [comment]
If you use the option `argsIgnorePattern: '^_.+'` instead of `varsIgnorePattern: '^_.+'`, it works as expected. It seems like ESlint might be treating everything to the left of the arrow function as parameters?

**Comment by phyllisstein** [comment]
Ah, I see now. This is arguably just a case of me misunderstanding the documentation, then. A destructured argument object's output _feels_ like a variable to me, but I'd be willing to accept it as an argument if that's its lifestyle choice.

**Comment by mysticatea** [comment]
Closing since I confirmed this is intentional behavior on [the demo](https://eslint.org/demo/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW51c2VkLXZhcnM6IFtlcnJvciwge1xyXG4gICAgYXJnczogYWZ0ZXItdXNlZCxcclxuICAgIGFyZ3NJZ25vcmVQYXR0ZXJuOiAnXl8uKycsXHJcbiAgICBpZ25vcmVSZXN0U2libGluZ3M6IGZhbHNlLFxyXG4gICAgdmFyczogYWxsLFxyXG4gICAgdmFyc0lnbm9yZVBhdHRlcm46ICdeXy4rJ1xyXG59XSAqL1xyXG5cclxuY29uc3QgZnVuID0gKHsgZm9vLCBiYXI6IF9iYXIgfSkgPT4gY29uc29sZS5sb2coZm9vKVxyXG5mdW4oKVxyXG4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjgsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=). Yes, `argsIgnorePattern` covers parameters.
Thank you.
